Integrate YOLO model

sjay05 opened the pull request:
- Added command line option to run script with YOLO or finding largest contour
- Uses 2023 CV model

sjay05 committed:
add 2023 YOLO landing-pad model

sjay05 committed:
integrate yolo decision

sjay05 committed:
add changes from black formatter

sjay05 committed:
edit argument field

TongguangZhang commented on blue_only.py:
```diff
@@ -166,49 +193,59 @@ def main() -> int:
-
             if largest_contour is not None:
-                # Draw a rectangle around the largest circular contour
                 x, y, w, h = cv2.boundingRect(largest_contour)
```
let's change this to early exiting like above
```
if largest_contour is None:
    continue
```

TongguangZhang commented on blue_only.py:
```diff
@@ -166,49 +193,59 @@ def main() -> int:
+                    x, y = best_landing_pad.x_1, best_landing_pad.y_1
+                    w = best_landing_pad.x_2 - best_landing_pad.x_1
+                    h = best_landing_pad.y_2 - best_landing_pad.y_1
+                    found_coordindates = True
```
let's change this to early exiting like above
```
if largest_contour is None:
    continue
```

TongguangZhang commented on blue_only.py:
```diff
@@ -166,49 +193,59 @@ def main() -> int:
+                    h = best_landing_pad.y_2 - best_landing_pad.y_1
+                    found_coordindates = True
+
+        if found_coordindates:
```
remove this if condition (since we want to have early exiting above)

TongguangZhang commented on models/best-2n.pt:
we try to not have models in the repositories (since their file size is really large) - let's add .pt files to the gitignore

sjay05 replied:
Removed.

TongguangZhang commented on yolo_decision.py:
```diff
@@ -0,0 +1,119 @@
+            label = int(boxes.cls[i])
+            confidence = float(boxes.conf[i])
+            result, detection = Detection.create(bounds, label, confidence)
+            if result and detection:
```
remember to return nothing if the create method fails (similar to the bootcamp)

sjay05 replied:
So if we have one good detection and one failed one, is it better to return nothing and continue? My original thought was it would be okay for the drone to land at the best landing pad among the valid detections.

sjay05 replied:
For now, I have edited it to return nothing if it fails.

TongguangZhang commented on blue_only.py:
```diff
@@ -123,6 +137,10 @@ def main() -> int:
     last_time = current_milli_time()
     last_image_time = current_milli_time()
+
+    if args.method == "yolo":
+        yolo_model = yolo_decision.DetectLandingPad("cpu", 0.5, "models/best-2n.pt")
```
move this to the section where you use the yolo model

TongguangZhang replied:
let's add these parameters to a config file (look at config.yaml in computer-vision-python for an example)

TongguangZhang commented on blue_only.py:
```diff
@@ -166,49 +193,59 @@ def main() -> int:
-                last_image_time = current_milli_time()
-            # break
+                found_coordindates = True
+        else:
```
to be explicit let's make this
`elif args.method == "yolo"`

TongguangZhang reviewed (commented):
Reviewed. Let's change it to have 2 functions - contour and yolo and main calls the function depending on the argument

sjay05 committed:
add .pt to gitignore

sjay05 committed:
refactor blue_only.py and add config.yml

sjay05 committed:
add black changes

sjay05 committed:
move parser to function

sjay05 commented:
Updated. 
- Two separate functions `run_with_countours()` or `yolo_detect` (run with contours is identical to before)
- Added YOLO model parameters to `config.yml`

TongguangZhang commented on blue_only.py:
```diff
@@ -213,6 +221,165 @@ def main() -> int:
+    )
+    args = parser.parse_args()
+
+    if args.method == "contour":
+        return run_with_contours()
+
+    if args.method == "yolo":
+        model_device = config["yolo_detect_target"]["device"]
+        detect_confidence = config["yolo_detect_target"]["confidence"]
+        model_path = config["yolo_detect_target"]["model_path"]
+
+        return yolo_detect(model_device, detect_confidence, model_path)
```
sorry I didn't mean to copy the entire script into the functions - we want to have the while loop in main, send the mavlink in main, etc. but just get the x, y, w, h from the contour or yolo function depending on the argparse. I was thinking more like - the line numbers are from the main branch and this isn't exact but like architecturally this is what I was thinking
```
def detect_landing_pads_contour():
    # basically copy paste lines 148-173
    return x, y, w, h

def detect_landing_pads_yolo():
    # implement the class and functions in yolo_detection.py to get a landing pad
    x, y, w, h = ... # interpret the results
    return x, y, w, h

def main():
    # setup stuff and argparse
    while True:
        # lines 87 - 148

          if args.method == "contour":
             x, y, w, h = detect_landing_pads_contour()
          if args.method == "yolo":
              x, y, w, h = detect_landing_pads_yolo()
          
          if x, y, w, h is None: # not exactly this but basically check that we have a bounding box
              return
          
          # lines 174 onward
```

TongguangZhang reviewed (commented):
Mostly revert to the old architecture - we just want the step of determining the landing pad to be in a function, not the whole setup and loop and mavlink message sending.

sjay05 committed:
modify blue_only

TongguangZhang commented on blue_only.py:
```diff
@@ -1,221 +1,290 @@
+    return message_id == pymavlink.mavutil.mavlink.MAVLINK_MSG_ID_RADIO_STATUS
+
+
+def detect_landing_pads_contour(image: np.ndarray) -> "None | tuple":
```
let's change the function to return `False, None` and `True, output`

TongguangZhang commented on blue_only.py:
```diff
@@ -1,221 +1,290 @@
+    return tuple(cv2.boundingRect(largest_contour))
+
+
+def detect_landing_pads_yolo(image: np.ndarray, config: dict) -> "None | tuple":
```
same as above, let's change the function to return `False, None` and `True, output`

TongguangZhang commented on blue_only.py:
```diff
@@ -1,221 +1,290 @@
+            print("No GLOBAL_POSITION_INT message received")
+            continue
+
+        if args.method == "contour":
+            bounding_box = detect_landing_pads_contour(image)
+        elif args.method == "yolo":
+            bounding_box = detect_landing_pads_yolo(image, config)
+
+        if bounding_box is not None:
```
from above, let's do `result, bounding_box = detect(image)` and check the result instead of the bounding box to see if it's valid

TongguangZhang replied:
let's do early exiting here - instead of `if result`, let's do
```
if not result:
    continue
```

TongguangZhang commented on blue_only.py:
```diff
@@ -1,221 +1,290 @@
+            loop_counter = 0
+            last_time = current_milli_time()
+        if current_milli_time() - last_image_time > 100:
+            if is_pad_detected:
```
same as above, instead of checking this here, have it exit early so we don't have to use a variable here

TongguangZhang reviewed (commented):
Reviewed

sjay05 committed:
add early exiting

TongguangZhang reviewed (approved):
LGTM
